### ISSUE TITLE
Using pretrained ELMo with TorchText 

### ISSUE THREAD

**Issue by littleflow3r** [bentrevett/pytorch-sentiment-analysis#25]
Hi, thanks for the tutorial. I have one question, though.
I want to use ELMo instead of GloVe or other for embedding, and suppose I already have ELMo representation for every sentence in the shape (seq_len, elmo_dimension). 
I want to either: 
1) concat this representation with the embedding from embedding layer or, 
2) use this representation before passing it to RNN/CNN.
Do you have any idea on how to use this with Torchtext? I am not sure how to add the elmo sentence representation to the batch and pass it to my model together with the input (which has been converted to indices) ?

Any advice/pointer would be greatly appreciated.

Thanks.

**Comment by bentrevett** [comment]
I haven't experimented with ELMo myself, but let's say you have your function that calculates ELMo embeddings on a batch of sentences called `get_elmo_embeddings`. Your training loop will look something like:

```python
def train(model, iterator, optimizer, criterion):
    
    epoch_loss = 0
    epoch_acc = 0
    
    model.train()
    
    for batch in iterator:
        
        optimizer.zero_grad()
        
        #batch.text = [seq len, batch]

        #get elmo embeddings for batch of sentences
        embedded = get_elmo_embeddings(batch.text)

        #embedded = [seq len, batch, elmo embedding dim]
 
        #use embeddings as input to model
        predictions = model(embedded).squeeze(1)
        
        loss = criterion(predictions, batch.label)
        
        acc = binary_accuracy(predictions, batch.label)
        
        loss.backward()
        
        optimizer.step()
        
        epoch_loss += loss.item()
        epoch_acc += acc.item()
        
    return epoch_loss / len(iterator), epoch_acc / len(iterator)
```
Then, within the model (using the CNN one as an example), you can get rid of the embedding layer as the words have already been converted into ELMo embeddings:

```python
class CNN(nn.Module):
    def __init__(self, vocab_size, elmo_embedding_dim, n_filters, filter_sizes, output_dim, dropout):
        super().__init__()
        
        #removed embedding layer as already have elmo embeddings

        self.convs = nn.ModuleList([
                                    nn.Conv2d(in_channels = 1, out_channels = n_filters, 
                                              kernel_size = (fs, elmo_embedding_dim)) 
                                    for fs in filter_sizes
                                    ])
        self.fc = nn.Linear(len(filter_sizes) * n_filters, output_dim)
        self.dropout = nn.Dropout(dropout)
        
    def forward(self, embedded):
        
        #words are already elmo embedded, so no need to embed again        

        #embedded = [sent len, batch size, elmo embedding dim]
        
        embedded = embedded.permute(1, 0)
                
        #embedded = [batch size, sent len, elmo embedding dim]
        
        embedded = embedded.unsqueeze(1)
        
        #embedded = [batch size, 1, sent len, elmo embedding dim]
        
        conved = [F.relu(conv(embedded)).squeeze(3) for conv in self.convs]
            
        #conv_n = [batch size, n_filters, sent len - filter_sizes[n]]
        
        pooled = [F.max_pool1d(conv, conv.shape[2]).squeeze(2) for conv in conved]
        
        #pooled_n = [batch size, n_filters]
        
        cat = self.dropout(torch.cat(pooled, dim=1))

        #cat = [batch size, n_filters * len(filter_sizes)]
            
        return self.fc(cat)
```

This is just replacing the learned embeddings with the ELMo embeddings, which is the way to go. I guess the other way would be to keep the embedding layer within the CNN module and concatenate it together with the ELMo embeddings making sure to have your kernel sizes be `embedding_dim + elmo_embedding_dim`.

**Comment by littleflow3r** [comment]
Hi, thanks for the reply! Actually I did exactly that, but I was not sure in this part:
` embedded = get_elmo_embeddings(batch.text) ` 
the batch.text here is already converted into indices, and to get ELMo representation we need original sentences (or tokenized). 
So what I did was converting the indices back to words to get the ELMo representation. It worked, but it seems inefficient (?) 
I wonder if it is possible to do like this with TorchText:
-read the data and turn into panda dataframe/list/dict
-get the elmo representation and save it with the df/list/dict
-read and split the data from the df/list/dict and build the vocab
-do the batching
What do you think?

**Comment by bentrevett** [comment]
Sorry, I misunderstood your original question. 
The only methods I can think of are doing it the way you have said was inefficient (transforming indexes back into words on the fly with `TEXT.vocab.itos`) or by doing something like your second idea, which would involve making a separate script that gets the ELMo embedding for each sentence and then writes the sentence, label and ELMo embeddings to a file which you can then read with TorchText instead of using `datasets.IMDB.splits(TEXT, LABEL)`.

**Comment by littleflow3r** [comment]
Thanks!
